seapath_ros_driver/src/seapath_ros_driver.cpp

```diff
@@ -184,55 +199,67 @@ namespace seapath
         kv6.key = "Heading Error";
         kv6.value = std::to_string(data.heading_error);
 
-        diagnostic_msg.values = {kv1, kv2, kv3, kv4, kv5, kv6};
+        auto flags = data.evaluateDiagnosticStatus();
 
-        if (position_error)
-        {
-            diagnostic_msg.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
-            diagnostic_msg.message = "Position error";
-        }
-        else if (orientation_error)
+
+        kv7.key = "Horizontal position and velocity status";
+        kv7.value = KMBinaryData::status_to_string(flags.horizontal_pos_vel);
+
+        kv8.key = "Roll and pitch status";
+        kv8.value = KMBinaryData::status_to_string(flags.roll_pitch);
+        
+        kv9.key = "Heading status";
+        kv9.value = KMBinaryData::status_to_string(flags.heading);
+
+        kv10.key = "Heave and vertical velocity status";
+        kv10.value = KMBinaryData::status_to_string(flags.heave_vert_vel);
+
+        kv11.key = "Acceleration status";
+        kv11.value = KMBinaryData::status_to_string(flags.acceleration);
+
+        kv12.key = "Delayed heave status";
+        kv12.value = KMBinaryData::status_to_string(flags.delayed_heave);
+
+        diagnostic_msg.values = {kv1, kv2, kv3, kv4, kv5, kv6, kv7, kv8, kv9, kv10, kv11, kv12};
+
+        // Determine the level and message based on error checks
+        switch (data.determineOverallStatus())
         {
+        case KMBinaryData::DiagnosticStatus::INVALID:
+            diagnostic_msg.level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
+            diagnostic_msg.message = "Invalid data";
+            break;
+        case KMBinaryData::DiagnosticStatus::REDUCED:
             diagnostic_msg.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
-            diagnostic_msg.message = "Orientation error";
-        }
-        else
-        {
+            diagnostic_msg.message = "Reduced performance";
+            break;
+        case KMBinaryData::DiagnosticStatus::OK:
             diagnostic_msg.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
-            diagnostic_msg.message = "Socket connection is OK";
+            diagnostic_msg.message = "All systems nominal";
+            break;
         }
-    
+        
         diagnostic_array.status.push_back(diagnostic_msg);
         return diagnostic_array;
     }
 
-    nav_msgs::msg::Odometry Driver::get_odometry_message(const KMBinaryData &data)
+    nav_msgs::msg::Odometry Driver::get_odometry_message(const KMBinaryData& data, const rclcpp::Time& time) const
```

Comment:
What is the motivation for not using the timestamp included in KMBinary?

Reply:
The time from the seapath is not synced with the pi and/or xavier. If Theo or someone else manages to sync the seapath with the rest of the system we can start using the timestamp from the KMBinary. The reason we can't do it now, although the time difference between subsequent KMBinary messages would be more accurate, tf2 functions will fail when the time difference is too big, as a result of the time not being synced. So this is the easiest method as of now.